### PR TITLE
✨🐛RE_3:Add support for CAaja ads network

### PR DESCRIPTION
**Pull request by matsuda-takehiro**
Adds support for CAaja ads network.

Hello.
We are A.J.A., a subsidiary of CyberAgent which provides ad-network service in Japan. (https://www.cyberagent.co.jp/) CA
(https://aja-kk.co.jp/) AJA
We'd like to add support to amp-ad.

Thanks.

previous↓
#21067

### CODE REVIEW

File: ads/aja.js [hunk 1]
```diff
@@ -14,28 +14,18 @@
...
   (global._aja = global._aja || {
-    viewId: global.context.pageViewId,
-    widgetIds: data['widgetids'],
-    htmlURL: data['htmlurl'] || global.context.canonicalUrl,
-    ampURL: data['ampurl'] || global.context.sourceUrl,
-    fbk: data['fbk'] || '',
-    testMode: data['testmode'] || 'false',
-    styleFile: data['stylefile'] || '',
-    referrer: data['referrer'] || global.context.referrer,
+    sspCode: data['sspCode'],
```
Comment (zhouyx) [inline ads/aja.js]: I've agreed on this https://github.com/ampproject/amphtml/pull/20652#issuecomment-465390283, but I want to bring this up one more time, just to be safe.  This is a breaking change for existing usage of `aja`, and we are good with it.

File: ads/aja.js [hunk 2]
```diff
@@ -14,28 +14,18 @@
...
+import {writeScript} from '../3p/3p';
 
 /**
  * @param {!Window} global
  * @param {!Object} data
  */
 export function aja(global, data) {
 
-  // ensure we have valid widgetIds value
-  validateData(data, ['widgetids']);
```
Comment (zhouyx) [inline ads/aja.js]: If `data-ssp-code` is a required param, you may still want to validate data.

File: ads/aja.js (hunk 1 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: really sorry. I do not seem to understand the meaning...

Would you like to open it again?
What should I do?

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

About 3.
What is the merit of <amp-embed>?
What happens if it gets removed from whitelist?

File: ads/aja.js (hunk 2 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: validate data...

Is there a check as to whether there is?

sorry..

File: ads/aja.js (hunk 1 above)
Comment (zhouyx) [reply ads/aja.js]: That the existing usage of `<amp-embed type='aja'>` will be broken. Do you have customers that embed aja ad in AMP pages today? And how do you want to support them after the change.

File: ads/aja.js (hunk 2 above)
Comment (zhouyx) [reply ads/aja.js]: Using the `#validateData()` helper method. 
`validateData(data, ['sspCode'])`, depending on if the `data-ssp-code` is required or not

**Comment by zhouyx** [comment]
If we remove it from whitelist `<amp-embed type='aja'>` will no longer be available, only `<amp-ad type='aja'>` works.

File: ads/aja.js (hunk 1 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: We have not yet delivered ads using <amp-embed type = 'aja'>....

I want to use <amp-ad> after this PR got through!

File: ads/aja.js (hunk 2 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: If not, is it a function for announcing?

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx


I do not understand a little yet, but I will change it a bit more.

**Comment by matsuda-takehiro** [comment]
```
  'aja': {
    renderStartImplemented: true,
    prefetch: 'https://cdn.as.amanad.adtdp.com/sdk/asot-v2.js',
  },
```
I do not really know what to put in prefetch. I'm sorry

File: ads/aja.js (hunk 1 above)
Comment (zhouyx) [reply ads/aja.js]: Got it! Thank you for confirming.

File: ads/aja.js (hunk 2 above)
Comment (zhouyx) [reply ads/aja.js]: How do you want to handle if there's no `ssp-code`? What `validateData` does is simply checking the existence of `ssp-code` and throw a user error if it's not there. I guess if you want some kind of fallback, you either check with `if/else` or use `try/catch`

File: ads/aja.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/aja.js]: This is fine, don't need to change the year for existing file, you can revert the change here.

File: ads/aja.js [hunk 4]
```diff
@@ -14,28 +14,18 @@
...
-    testMode: data['testmode'] || 'false',
-    styleFile: data['stylefile'] || '',
-    referrer: data['referrer'] || global.context.referrer,
+    sspCode: data['sspCode'],
   });
 
-  // load the Aja AMP JS file
-  loadScript(global, 'https://cdn.as.amanad.adtdp.com/sdk/asot-v2.js');
+  const url = 'https://static.aja-recommend.com/html/amp.html?ssp_code=' + encodeURIComponent(data['sspCode']);
+  writeScript(global,url);
```
Comment (zhouyx) [inline ads/aja.js]: You are adding a html file here in side a script tag. I don't believe it functions properly. Please double check. Thank you

File: ads/aja.md [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <!---
-Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+Copyright 2019 The AMP HTML Authors. All Rights Reserved.
```
Comment (zhouyx) [inline ads/aja.md]: ditto: same, please revert year change.

File: ads/aja.md [hunk 6]
```diff
@@ -21,29 +21,20 @@ limitations under the License.
 ### Basic
 
 ```html
-  <amp-embed width="100" height="100"
+  <amp-ad width="100" height="100"
              type="aja"
              layout="responsive"
-             data-widgetIds="AMP_1,AMP_2">
-  </amp-embed>
+             data-media-id="ZZZZZ">
```
Comment (zhouyx) [inline ads/aja.md]: Is `data-media-id` also a required param?

**Comment by zhouyx** [comment]
@matsuda-takehiro Please refer to the doc [here](https://github.com/ampproject/amphtml/blob/master/ads/README.md#preconnect-and-prefetch) for prefetch and preconnect setting. 
You can leave the config empty if there's nothing to preconnect/prefetch.

File: ads/aja.js (hunk 2 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: understood!
I will add it on weekdays!

File: ads/aja.js (hunk 4 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: What if I want to use html files?

File: ads/aja.md (hunk 5 above)
Comment (matsuda-takehiro) [reply ads/aja.md]: We will fix it on weekdays

File: ads/aja.js (hunk 3 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: We will fix it on weekdays

File: ads/aja.md (hunk 6 above)
Comment (matsuda-takehiro) [reply ads/aja.md]: I made a mistake

**Comment by matsuda-takehiro** [comment]
renderStartImplemented: true

Do you want to leave this string alone?

File: ads/aja.js (hunk 3 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: fix it!

File: ads/aja.md (hunk 5 above)
Comment (matsuda-takehiro) [reply ads/aja.md]: fix it!

File: ads/aja.js (hunk 2 above)
Comment (matsuda-takehiro) [reply ads/aja.js]: An error occurred😭

Comment (matsuda-takehiro) [reply ads/aja.js]: image

...?

Comment (matsuda-takehiro) [reply ads/aja.js]: maybe fix it

**Comment by matsuda-takehiro** [comment]
Is anything displayed?

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

It will be summarized as follows
- renderStartImplemented: Do you want to keep true?
- Do you run the test and see something?
- What if I want to load an html file instead of js?

**Comment by zhouyx** [comment]
@matsuda-takehiro Thank you for the summary.

>renderStartImplemented: Do you want to keep true?

We recommend implementing the `renderStart` API for better user experience. So please implement it if possible. You can refer to https://github.com/ampproject/amphtml/blob/master/ads/README.md#available-apis

>Do you run the test and see something?

I couldn't see an ad loading because of the third point. 

>What if I want to load an html file instead of js?

This will be exactly the same with how you load html file on non AMP pages. I think you can create another iframe, and set its `src`.  This is an example https://github.com/ampproject/amphtml/blob/master/ads/medyanet.js

**Comment by matsuda-takehiro** [comment]
Thank you very much for helping me　@zhouyx

Is it okay to issue out using ```document.createElement```? I will try it at once!

**Comment by matsuda-takehiro** [comment]
What is percy?

**Comment by matsuda-takehiro** [comment]
elStyle.setAttribute('src', 'https://static.aja-recommend.com/html/amp.html?ssp_code=' + encodeURIComponent(data['sspCode']));

🤔

**Comment by matsuda-takehiro** [comment]
In a pattern using CDN tag

Using SSPID and MEDIAID, I thought of generating tags like
<script src = "https://cdn.as.amanad.adtdp.com/tag/MEDIAID.js" data - aja - ssp - asi = "SSPID" async = "true"> </ script>

**Comment by zhouyx** [comment]
> Is it okay to issue out using document.createElement?

Could you explain further? 

>What is percy?

Percy is the tool we use for visual diff. Your change should not lead to any visual diff change, so feel free to ignore the failure there. 

>In a pattern using CDN tag
Using SSPid and MEDIAid, I thought of generating tags like
<script src = "https://cdn.as.amanad.adtdp.com/tag/MEDIAID.js" data - aja - ssp - asi = "SSPID" async = "true">

Could you please further explain or is this a question? Thank you

**Comment by matsuda-takehiro** [comment]
Thank you @ zhouyx

1.  please do not worry. It is a mistake of my writing

2. Thank you!

3. If I do not see it in the current change (loading amp-html), I thought I wanted to show the CDN tag. In that case, not only SSPCODE but MEDIA ID is required. I will add at that time.


Do you see something now?

**Comment by zhouyx** [comment]
> If I do not see it in the current change (loading amp-html), I thought I wanted to show the CDN tag. In that case, not only SSPCODE but MEDIA ID is required. I will add at that time.

Adding support to `mediaId` will be very similar to `sspCode`.

File: ads/aja.md [hunk 7]
```diff
@@ -21,29 +21,20 @@ limitations under the License.
...
 
-### Widget is cut off
-
-According to the AMP API, "resizes are honored when the resize will not adjust the content the user is currently reading.  That is, if the ad is above the viewport's contents, it'll resize. Same if it's below. If it's in the viewport, it ignores it."
-
-**Resolution**
-
- You can set an initial height of what the widget height is supposed to be. That is, instead of ```height="100"```, if the widget's final height is 600px, then set ```height="600"```. Setting the initial height ***will not*** finalize the widget height if it's different from the actual. The widget will resize to it's true dimensions after the widget leaves the viewport.
+- SSPCODE
+  - `data-aja-ssp-asi`
```
Comment (zhouyx) [inline ads/aja.md]: I don't see `data-aja-ssp-asi`? Can we update the doc

**Comment by zhouyx** [comment]
@matsuda-takehiro Thank you I am able to see the ad now.
 One thing I notice is that no `render-start` message has been received. Could you please verify that that renderStart API is implemented correctly. Thank you

**Comment by matsuda-takehiro** [comment]
I'm sorry, give me a screenshot!

**Comment by matsuda-takehiro** [comment]
Thank you @ zhouyx


I did not understand how to perfectly implement ```render-start``` ...

File: ads/aja.md (hunk 7 above)
Comment (matsuda-takehiro) [reply ads/aja.md]: sorry, I'll fix it now!

**Comment by zhouyx** [comment]
'render-start' and 'no-fill' are simply two postMessages that tells the AMP page to remove placeholder and display the ad, or try to collapse the ad slot. Maybe you can refer to other ad networks' implementation to these two APIs?

File: ads/aja.md (hunk 7 above)
Comment (matsuda-takehiro) [reply ads/aja.md]: It was data - ssp - code, not data - aja - ssp - asi! It seems they were mistaken for CDN tags. I'm sorry

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

1. 'render-start'
I could not find anything in common with Ad Network.js ...
really sorry

2.Do you see the image of some advertisements?
I am very happy! give me a screenshot because I want to see myself too?

Thank you so much for being with me for a long time!

**Comment by zhouyx** [comment]
1. `render-start`, it really depends on when do you think it's the best time to send render-start signal. One thing you can do is to call `window.context.renderStart();` at the iframe onload event. I'm pretty sure that the previous example I showed you, the medyanet did the same thing.

2. For sure, I can send u a screenshot later. Currently working with different branch.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

1.```window.context.renderStart(); at the iframe onload event. ```
I see! Very helpful in descriptive explanation!
However, I do not have my own technical skills and how to implement it is unknown.
Is it OK even if not implemented this time?

2.I am very sorry.
I'm sorry during working on another branch. ! Thank you!

**Comment by zhouyx** [comment]
I would highly recommend supporting that. But it's your final call. I'm fine with that.

**Comment by matsuda-takehiro** [comment]
```render-start```
In the future, I will make top priority when I get used!（If technical skill improves）

**Comment by zhouyx** [comment]
Here's the screenshot
image

**Comment by zhouyx** [comment]
SG, feel free to remove `renderStartImplemented` from your config.

**Comment by zhouyx** [comment]
And as agreed can we have @hideaki0817's approval on the change. Thank you

**Comment by matsuda-takehiro** [comment]
I asked him to give me a comment by hideaki!
Thank you very much for today!

**Comment by matsuda-takehiro** [comment]
I am planning to receive approval comments today

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx
I am planning to receive approval comments today

**Comment by hideaki0817** [comment]
I proved.

**Comment by zhouyx** [comment]
Thank you @matsuda-takehiro and @hideaki0817. The change has been merged, you can expect the change to hit prod in 2 weeks.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx 
I apologize for the inconvenience this may cause you.

Looking forward to two weeks applied!

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

Thank you very much for the last week!
Did you release it?

I have not confirmed the update ...

**Comment by matsuda-takehiro** [comment]
It was captured! Thank you!

**Comment by zhouyx** [comment]
Great! Glad it works
